pkg/api/status/v1.go

```diff
@@ -0,0 +1,126 @@
+// Copyright (c) The Thanos Authors.
+// Licensed under the Apache License 2.0.
+
+// Copyright 2016 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package status
+
+import (
+	"fmt"
+	"math"
+	"net/http"
+
+	"github.com/thanos-io/thanos/pkg/api"
+
+	"github.com/go-kit/log"
+	"github.com/opentracing/opentracing-go"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/common/route"
+	"github.com/prometheus/prometheus/model/labels"
+	"github.com/prometheus/prometheus/tsdb"
+	"github.com/prometheus/prometheus/tsdb/index"
+	v1 "github.com/prometheus/prometheus/web/api/v1"
+	extpromhttp "github.com/thanos-io/thanos/pkg/extprom/http"
+	"github.com/thanos-io/thanos/pkg/logging"
+)
+
+// Stat holds the information about individual cardinality.
+type Stat struct {
+	Name  string `json:"name"`
+	Value uint64 `json:"value"`
+}
+
+func convertStats(stats []index.Stat) []Stat {
+	result := make([]Stat, 0, len(stats))
+	for _, item := range stats {
+		item := Stat{Name: item.Name, Value: item.Count}
+		result = append(result, item)
+	}
+	return result
+}
+
+// TSDBStatus has information of cardinality statistics from postings.
+type TSDBStatus struct {
+	HeadStats                   v1.HeadStats `json:"headStats"`
+	SeriesCountByMetricName     []Stat       `json:"seriesCountByMetricName"`
+	LabelValueCountByLabelName  []Stat       `json:"labelValueCountByLabelName"`
+	MemoryInBytesByLabelName    []Stat       `json:"memoryInBytesByLabelName"`
+	SeriesCountByLabelValuePair []Stat       `json:"seriesCountByLabelValuePair"`
+}
+
+type GetStatsFunc func(r *http.Request, statsByLabelName string) (*tsdb.Stats, error)
+
+type Options struct {
+	GetStats GetStatsFunc
+	Registry *prometheus.Registry
+}
+
+type StatusAPI struct {
+	getTSDBStats GetStatsFunc
+	options      Options
+}
+
+func New(opts Options) *StatusAPI {
+	return &StatusAPI{
+		getTSDBStats: opts.GetStats,
+		options:      opts,
+	}
+}
+
+func (sapi *StatusAPI) Register(r *route.Router, tracer opentracing.Tracer, logger log.Logger, ins extpromhttp.InstrumentationMiddleware, logMiddleware *logging.HTTPServerMiddleware) {
+	instr := api.GetInstr(tracer, logger, ins, logMiddleware, false)
+	r.Get("/api/v1/status/tsdb", instr("tsdb_status", sapi.httpServeStats))
+}
+
+func (sapi *StatusAPI) httpServeStats(r *http.Request) (interface{}, []error, *api.ApiError) {
+	s, err := sapi.getTSDBStats(r, labels.MetricName)
```

Comment:
So this API can only fetch the TSDB stats for a specific tenant. Admins cannot use this API to to see the global TSDB stats (cross tenant).

Reply:
Yes, that's correct. We could extend the endpoint to provide a global view by traversing all tenants and merging the results.

Reply:
Sounds good. Let's have an issue to track it after this pr is merged.